### PR TITLE
4.0 (Tippy 6)

### PR DESCRIPTION
**Pull request by atomiks**
## Breaking Changes

Aside from `tippy.js@6` changes:

- Deleted `<TippySingleton />`
- Change `enabled` to `disabled`
- `useSingleton()` API changed to support headless rendering

## Headless Tippy using `render` prop

Better CSS-in-JS integration, `react-spring` integration, & more!

```jsx
<Tippy render={attrs => <div className="box" {...attrs}>Hello</div>}>
  <button>Hello</button>
</Tippy>
```

Thoughts @KubaJastrz?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -67,7 +67,7 @@
     "@types/react": "^16.8.2",
     "babel-core": "^7.0.0-bridge.0",
     "babel-eslint": "^10.0.1",
-    "babel-jest": "^23.6.0",
+    "babel-jest": "^25.1.0",
```
Comment (KubaJastrz) [inline package.json]: does it work without upgrading `jest` to `^25.1.0`?

**Comment by atomiks** [comment]
> I don't think I understand the template prop. In your example it just takes precedence over content prop and conditionally sets render tippy prop to some constant value.

Right, if it's `undefined`, it will fallback to the `content` prop. It's constant because React does the DOM diffing for us, so we just need to pass the ref. 

The conditional checks allow us to keep 1 component that we can use `TippyGenerator()` to inject the `tippy()` core constructor into instead of duplicate files.

**Comment by KubaJastrz** [comment]
Ok, so why would I ever want to use `template` prop then?

**Comment by atomiks** [comment]
> Ok, so why would I ever want to use template prop then?

CSS in JS without globals, react-spring animation integration, etc

**Comment by KubaJastrz** [comment]
How I see this:

1. Keep `content` prop as it is now.
2. Add new `render` prop that is a function that receives passed content (or even whole instance) and outputs `box` element. It would work the same as the `render` prop  in [headless tippy](https://atomiks.github.io/tippyjs/v6/headless-tippy/) example.

This allows for easy migration between built-in and user-provided template. I think this also suits the React render prop pattern better.

An example can be found in how [`selectivity` handled it](https://github.com/arendjr/selectivity/blob/ad2d158659de4bd73241dd4b19d9b68481f8ad62/src/templates.js) some time ago.

Also I think `<Tippy template={<div className="box">Hello</div>} />` is not very descriptive (if `template` were to be interchangeable with `content`) and could lead to confusion.

**Comment by atomiks** [comment]
The reason I wasn't going to make it a `render` prop function is because it doesn't really make sense. Headless Tippy works differently to how Headless React libraries work. We don't need to pass anything because there's no useful information to pass. Usually, vanilla users access the `props` object, but with `template`, they already know the props passed in from a component

**Comment by KubaJastrz** [comment]
> We don't need to pass anything because there's no useful information to pass.

Ok, that makes sense. I was thinking of some sort of tippy wrapper components scenario but I guess they would know all the props passing through anyway.

Maybe it's just the terminology that seems off. It's not really a template as it doesn't accept any input, is it?

**Comment by atomiks** [comment]
> Maybe it's just the terminology that seems off. It's not really a template as it doesn't accept any input, is it?

From what I can tell, "template" means DOM structrue with some variables / information injected. I think that fits here? e.g. Tooltip.js was similar, but was a string instead of JSX

**Comment by atomiks** [comment]
OK there is some useful info 😅 - `data-placement`, `data-reference-hidden` etc attributes to be placed on the template/box node.  In Headless Tippy these are placed on the root popper node (default `applyStyles` modifier), but can't be used for styling.

File: index.d.ts [hunk 2]
```diff
@@ -1,30 +1,36 @@
...
   visible?: boolean;
-  enabled?: boolean;
+  disabled?: boolean;
   className?: string;
   singleton?: (instance: Instance) => void;
+  render?: (attrs: {
+    'data-placement': Placement;
+    'data-reference-hidden'?: string;
+    'data-escaped'?: string;
+  }) => React.ReactNode;
```
Comment (atomiks) [inline index.d.ts]: This allows for very easy spreading: `render={attrs => <Box {...attrs} />}`. If they want to rename the attributes to something else, that's obviously also still possible, but a little bit unclean with the way it's accessed, e.g. 

```jsx
<Tippy
  render={attrs => (
    <Box
      data-someprefix-placement={attrs['data-placement']}
      data-someprefix-reference-hidden={attrs['data-reference-hidden']}
    />
  )}
/>
```

**Comment by KubaJastrz** [comment]
How can I re-implement [all props with R](https://atomiks.github.io/tippyjs/v6/all-props/) now that `render` prop in `tippy` is overwritten?

**Comment by atomiks** [comment]
React users pass props in the component, rather than access props in the tippy instance. So for example `"theme"` can be passed as a prop to their component and can be implemented in any way they like.

```jsx
const Tooltip = ({theme, label}) => (
  <Tippy 
    content={label} 
    render={attrs => <Box data-theme={theme} {...attrs} />}
  >
    <button />
  </Tippy>
);
```

**Comment by KubaJastrz** [comment]
Ok, I understand it now. Users will do all the work in `attrs => <Box {...attrs} />` if needed.

**Comment by FuJuntao** [comment]
> ## Breaking Changes
> Aside from `tippy.js@6` changes:
> 
> * Deleted `<TippySingleton />`
> * Change `enabled` to `disabled`
> 
> ## Headless Tippy using `render` prop
> Better CSS-in-JS integration, `react-spring` integration, & more!
> 
> ```js-jsx
> <Tippy render={attrs => <div className="box" {...attrs}>Hello</div>}>
>   <button>Hello</button>
> </Tippy>
> ```
> 
> Thoughts @KubaJastrz?

So breaking changes are only these two, no more else right?

**Comment by atomiks** [comment]
@FuJuntao some core library changes were made as well (and Popper 2 upgrade), you can [read the migration guide here](https://github.com/atomiks/tippyjs/blob/master/MIGRATION_GUIDE.md)

**Comment by atomiks** [comment]
https://github.com/atomiks/tippy.js-react/releases/tag/v4.0.0-alpha.0

Feel free to play with it and find issues 🤗

**Comment by atomiks** [comment]
Hmmm.. I don't think we can `useSingleton` with Headless Tippy... it might need to be a component + render prop because we need to use `createPortal` 👀

**Comment by ckeeney** [comment]
When upgrading `@tippy.js/react` from `v2` to `v3`, text in the tooltips became blurry.  Since Tippy 6 uses Popper v2, I believe this problem will go away in `v4`.

Until then, disabling GPU acceleration in the popper options seems to fix the blurry text.

```js
<Tippy
    content={tooltipContent}
    popperOptions={{modifiers: {computeStyle: {gpuAcceleration: false}}}}
/>
```

Related: 
https://github.com/atomiks/tippyjs/issues/163
https://github.com/popperjs/popper-core/pull/879

File: src/forwardRef.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import React, {cloneElement, forwardRef} from 'react';
+import {preserveRef} from './utils';
+
+export default Tippy =>
```
Comment (KubaJastrz) [inline src/forwardRef.js]: maybe let's export a named function for better debugging?

Comment (atomiks) [reply src/forwardRef.js]: I don't think it matters in this case

**Comment by atomiks** [comment]
I'm merging this but not doing a final release yet, it'll still be alpha. The only way I can get people to "try this out" is to make it the default. I'll still want to make potential breaking changes before final release, but I need feedback to do so.

**Pull request merged**